File: sdk/js/smoke_test.mjs
#!/usr/bin/env node
// Smoke test for the Autopilon JS/TS SDK (Phase 17 §55) — a plain script
// against a REAL, already-booted server with a real API key, same style
// as server/test/*.js and sdk/python/smoke_test.py. Run with:
//
//   AUTOPILON_API_KEY=ap_live_... AUTOPILON_BASE_URL=http://localhost:4000/api/v1 \
//       node smoke_test.mjs
//
// Exits non-zero if any check fails.
import { AutopilonClient, AutopilonApiError } from "./index.js";

const BASE_URL = process.env.AUTOPILON_BASE_URL || "http://localhost:4000/api/v1";
const API_KEY = process.env.AUTOPILON_API_KEY;

if (!API_KEY) {
  console.error("AUTOPILON_API_KEY is required (export it or pass inline).");
  process.exit(1);
}

const client = new AutopilonClient({ apiKey: API_KEY, baseUrl: BASE_URL });
const results = [];

async function check(name, fn) {
  try {
    await fn();
    results.push({ name, ok: true });
    console.log(`PASS  ${name}`);
  } catch (err) {
    results.push({ name, ok: false, error: err.message });
    console.log(`FAIL  ${name}: ${err.message}`);
  }
}

await check("agents.list() returns a paginated shape", async () => {
  const page = await client.agents.list({ limit: 10 });
  if (!Array.isArray(page.data)) throw new Error("data is not a list");
  if (!("hasMore" in (page.pagination || {}))) throw new Error("missing pagination.hasMore");
});

await check("tasks create/get/update/complete/archive round-trip", async () => {
  const task = await client.tasks.create({ title: "JS SDK smoke test task", priority: "high" });
  if (task.title !== "JS SDK smoke test task") throw new Error("title mismatch on create");
  const fetched = await client.tasks.get(task.id);
  if (fetched.id !== task.id) throw new Error("id mismatch on get");
  const updated = await client.tasks.update(task.id, { description: "updated via js sdk" });
  if (updated.description !== "updated via js sdk") throw new Error("description not updated");
  const completed = await client.tasks.complete(task.id);
  if (completed.status !== "completed") throw new Error("status not completed");
  const archived = await client.tasks.archive(task.id);
  if (archived.status !== "archived") throw new Error("status not archived");
});

await check("agents.get(nonexistent) raises AutopilonApiError with RESOURCE_NOT_FOUND", async () => {
  try {
    await client.agents.get("does-not-exist");
    throw new Error("expected AutopilonApiError");
  } catch (err) {
    if (!(err instanceof AutopilonApiError)) throw err;
    if (err.code !== "RESOURCE_NOT_FOUND") throw new Error(`wrong code: ${err.code}`);
    if (err.status !== 404) throw new Error(`wrong status: ${err.status}`);
    if (!err.requestId) throw new Error("missing requestId");
  }
});

await check("webhooks create/getSecret/sendTestEvent/list/delete round-trip", async () => {
  const webhook = await client.webhooks.create({ url: "https://example.com/hooks/js-sdk-test", events: ["task.created"] });
  if (!webhook.secret?.startsWith("whsec_")) throw new Error("no secret returned on create");
  const secret = await client.webhooks.getSecret(webhook.id);
  if (secret.secret !== webhook.secret) throw new Error("secret mismatch on re-fetch");
  const listing = await client.webhooks.list();
  if (!listing.data.some((w) => w.id === webhook.id)) throw new Error("created webhook not in list");
  const testResult = await client.webhooks.sendTestEvent(webhook.id);
  if (typeof testResult.responseTimeMs !== "number") throw new Error("test event missing responseTimeMs");
  await client.webhooks.delete(webhook.id);
});

await check("projects.create requires a workspaceId (INVALID_REQUEST)", async () => {
  try {
    await client.projects.create({ workspaceId: null, name: "no workspace" });
    throw new Error("expected AutopilonApiError");
  } catch (err) {
    if (!(err instanceof AutopilonApiError)) throw err;
    if (err.code !== "INVALID_REQUEST") throw new Error(`wrong code: ${err.code}`);
  }
});

await check("marketplace.listAssets + listCategories return real shapes", async () => {
  const assets = await client.marketplace.listAssets({ limit: 5 });
  if (!Array.isArray(assets.data)) throw new Error("assets.data is not a list");
  const categories = await client.marketplace.listCategories();
  if (!Array.isArray(categories.data)) throw new Error("categories.data is not a list");
});

await check("files upload/get/downloadContent/delete round-trip", async () => {
  const content = "hello from the js sdk smoke test";
  const blob = new Blob([content], { type: "text/plain" });
  const uploaded = await client.files.upload({ file: blob, filename: "js-sdk-test.txt" });
  if (uploaded.filename !== "js-sdk-test.txt") throw new Error("filename mismatch");
  const fetched = await client.files.get(uploaded.id);
  if (fetched.id !== uploaded.id) throw new Error("id mismatch on get");
  const downloadRes = await client.files.downloadContent(uploaded.id);
  const downloaded = await downloadRes.text();
  if (downloaded !== content) throw new Error(`content mismatch: ${downloaded}`);
  await client.files.delete(uploaded.id);
});

await check("tasks.create with idempotencyKey replays on retry", async () => {
  const key = `js-sdk-smoke-${Date.now()}`;
  const first = await client.tasks.create({ title: "JS SDK idempotency test" }, { idempotencyKey: key });
  const second = await client.tasks.create({ title: "JS SDK idempotency test" }, { idempotencyKey: key });
  if (first.id !== second.id) throw new Error("expected the same task id on replay");
});

await check("integrations.listActions(unconnected provider) -> 404", async () => {
  try {
    await client.integrations.listActions("gmail");
    throw new Error("expected AutopilonApiError");
  } catch (err) {
    if (!(err instanceof AutopilonApiError)) throw err;
    if (err.status !== 404) throw new Error(`expected 404, got ${err.status}`);
  }
});

const passed = results.filter((r) => r.ok).length;
console.log(`\n${passed}/${results.length} checks passed.`);
process.exit(passed === results.length ? 0 : 1);

File: sdk/python/smoke_test.py
#!/usr/bin/env python3
"""Smoke test for the Autopilon Python SDK (Phase 17.1 §4) — a plain script
against a REAL, already-booted server with a real API key, same style as
server/test/*.js. No test framework required (this SDK has zero
dependencies, including for testing itself). Run with:

    AUTOPILON_API_KEY=ap_live_... AUTOPILON_BASE_URL=http://localhost:4000/api/v1 \\
        python3 smoke_test.py

Exits non-zero if any check fails.
"""
import os
import sys
import time

sys.path.insert(0, os.path.dirname(__file__))
from autopilon import AutopilonClient, AutopilonApiError  # noqa: E402

BASE_URL = os.environ.get("AUTOPILON_BASE_URL", "http://localhost:4000/api/v1")
API_KEY = os.environ.get("AUTOPILON_API_KEY")

if not API_KEY:
    print("AUTOPILON_API_KEY is required (export it or pass inline).")
    sys.exit(1)

client = AutopilonClient(api_key=API_KEY, base_url=BASE_URL)

results = []


def check(name, fn):
    try:
        fn()
        results.append((name, True, None))
        print(f"PASS  {name}")
    except Exception as err:  # noqa: BLE001 - intentionally broad, this is a test harness
        results.append((name, False, str(err)))
        print(f"FAIL  {name}: {err}")


def check_agents_list():
    page = client.agents.list(limit=10)
    assert isinstance(page.get("data"), list), "data is not a list"
    assert "hasMore" in page.get("pagination", {}), "missing pagination.hasMore"


def check_task_roundtrip():
    task = client.tasks.create(title="Python SDK smoke test task", priority="high")
    assert task["title"] == "Python SDK smoke test task"
    fetched = client.tasks.get(task["id"])
    assert fetched["id"] == task["id"]
    updated = client.tasks.update(task["id"], description="updated via python sdk")
    assert updated["description"] == "updated via python sdk"
    completed = client.tasks.complete(task["id"])
    assert completed["status"] == "completed"
    archived = client.tasks.archive(task["id"])
    assert archived["status"] == "archived"


def check_pagination():
    client.tasks.create(title="Second python sdk task")
    seen = []
    for item in client.paginate("/tasks", query={"limit": 1}):
        seen.append(item["id"])
        if len(seen) > 10:
            break
    assert len(seen) >= 2, f"expected >=2 tasks via pagination, got {len(seen)}"


def check_not_found_error_shape():
    try:
        client.agents.get("does-not-exist")
        raise AssertionError("expected AutopilonApiError")
    except AutopilonApiError as err:
        assert err.code == "RESOURCE_NOT_FOUND", f"wrong code: {err.code}"
        assert err.status == 404, f"wrong status: {err.status}"
        assert err.request_id, "missing request_id"


def check_webhook_roundtrip():
    webhook = client.webhooks.create(url="https://example.com/hooks/python-sdk-test", events=["task.created"])
    assert webhook["secret"].startswith("whsec_"), "no secret returned on create"
    secret = client.webhooks.get_secret(webhook["id"])
    assert secret["secret"] == webhook["secret"], "secret mismatch on re-fetch"
    listing = client.webhooks.list()
    assert any(w["id"] == webhook["id"] for w in listing["data"]), "created webhook not in list"
    test_result = client.webhooks.send_test_event(webhook["id"])
    assert isinstance(test_result.get("responseTimeMs"), int), "test event missing responseTimeMs"
    client.webhooks.delete(webhook["id"])


def check_project_validation_error():
    try:
        client.projects.create(workspace_id=None, name="no workspace")
        raise AssertionError("expected AutopilonApiError")
    except AutopilonApiError as err:
        assert err.code == "INVALID_REQUEST", f"wrong code: {err.code}"


def check_marketplace():
    assets = client.marketplace.list_assets(limit=5)
    assert isinstance(assets.get("data"), list)
    categories = client.marketplace.list_categories()
    assert isinstance(categories.get("data"), list)


def check_file_roundtrip():
    content = b"hello from the python sdk smoke test"
    uploaded = client.files.upload(content, filename="python-sdk-test.txt", content_type="text/plain")
    assert uploaded["filename"] == "python-sdk-test.txt"
    fetched = client.files.get(uploaded["id"])
    assert fetched["id"] == uploaded["id"]
    downloaded = client.files.download_content(uploaded["id"])
    assert downloaded == content, f"content mismatch: {downloaded!r}"
    client.files.delete(uploaded["id"])


def check_idempotency():
    key = f"python-sdk-smoke-{time.time()}"
    first = client.tasks.create(title="Python SDK idempotency test", idempotency_key=key)
    second = client.tasks.create(title="Python SDK idempotency test", idempotency_key=key)
    assert first["id"] == second["id"], "expected the same task id on replay"


def check_integration_actions_not_found():
    try:
        client.integrations.list_actions("gmail")
        raise AssertionError("expected AutopilonApiError")
    except AutopilonApiError as err:
        assert err.status == 404, f"expected 404 for an unconnected provider, got {err.status}"


check("agents.list() returns a paginated shape", check_agents_list)
check("tasks create/get/update/complete/archive round-trip", check_task_roundtrip)
check("client.paginate() walks tasks without manual cursors", check_pagination)
check("agents.get(nonexistent) raises AutopilonApiError with RESOURCE_NOT_FOUND", check_not_found_error_shape)
check("webhooks create/get_secret/send_test_event/list/delete round-trip", check_webhook_roundtrip)
check("projects.create requires a workspace_id (INVALID_REQUEST)", check_project_validation_error)
check("marketplace.list_assets + list_categories return real shapes", check_marketplace)
check("files upload/get/download_content/delete round-trip", check_file_roundtrip)
check("tasks.create with idempotency_key replays on retry", check_idempotency)
check("integrations.list_actions(unconnected provider) -> 404", check_integration_actions_not_found)

passed = sum(1 for _, ok, _ in results if ok)
print(f"\n{passed}/{len(results)} checks passed.")
sys.exit(0 if passed == len(results) else 1)
